nifi-nar-bundles/nifi-email-bundle/nifi-email-processors/src/main/java/org/apache/nifi/processors/email/smtp/handler/SMTPMessageHandlerFactory.java

```diff
@@ -0,0 +1,147 @@
+ * limitations under the License.
+ */
+
+package org.apache.nifi.processors.email.smtp.handler;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.cert.X509Certificate;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.stream.io.ByteArrayOutputStream;
+import org.apache.nifi.util.StopWatch;
+import org.subethamail.smtp.DropConnectionException;
+import org.subethamail.smtp.MessageContext;
+import org.subethamail.smtp.MessageHandler;
+import org.subethamail.smtp.MessageHandlerFactory;
+import org.subethamail.smtp.RejectException;
+import org.subethamail.smtp.TooMuchDataException;
+import org.subethamail.smtp.server.SMTPServer;
+
+import org.apache.nifi.processors.email.smtp.event.SmtpEvent;
+
+
+public class SMTPMessageHandlerFactory implements MessageHandlerFactory {
+    final LinkedBlockingQueue<SmtpEvent> incomingMessages;
+    final ComponentLog logger;
+
+    public SMTPMessageHandlerFactory(LinkedBlockingQueue<SmtpEvent> incomingMessages, ComponentLog logger) {
+        this.incomingMessages = incomingMessages;
+        this.logger = logger;
+    }
+
+    @Override
+    public MessageHandler create(MessageContext messageContext) {
+        return new Handler(messageContext, incomingMessages, logger);
+    }
+
+    class Handler implements MessageHandler {
+        final MessageContext messageContext;
+        String from;
+        String recipient;
+        byte [] messageBody;
+
+
+        public Handler(MessageContext messageContext, LinkedBlockingQueue<SmtpEvent> incomingMessages, ComponentLog logger){
+            this.messageContext = messageContext;
+        }
+
+        @Override
+        public void from(String from) throws RejectException {
+            // TODO: possibly whitelist senders?
+            this.from = from;
+        }
+
+        @Override
+        public void recipient(String recipient) throws RejectException {
+            // TODO: possibly whitelist receivers?
+            this.recipient = recipient;
+        }
+
+        @Override
+        public void data(InputStream inputStream) throws RejectException, TooMuchDataException, IOException {
+            // Start counting the timer...
+
+            StopWatch watch = new StopWatch(false);
+
+            SMTPServer server = messageContext.getSMTPServer();
+
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+            byte [] buffer = new byte[1024];
+            int rd;
+
+            while ((rd = inputStream.read(buffer, 0, buffer.length)) != -1) {
+                baos.write(buffer, 0, rd);
+            }
+            if (baos.getBufferLength() > server.getMaxMessageSize()) {
+                throw new TooMuchDataException("Data exceeds the amount allowed.");
+            }
+
+            baos.flush();
+            this.messageBody = baos.toByteArray();
```

Comment:
Now that you are processing the message in the "data" section, why not pass the "inputStream" itself instead of the message body to the ontrigger? That way you don't have to read the whole message into memory and can just write the data straight to the flowfile content.